scipy/optimize/zeros.py

```diff
@@ -112,36 +134,48 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     -----
     The convergence rate of the Newton-Raphson method is quadratic,
     the Halley method is cubic, and the secant method is
-    sub-quadratic.  This means that if the function is well behaved
-    the actual error in the estimated zero is approximately the square
-    (cube for Halley) of the requested tolerance up to roundoff
-    error. However, the stopping criterion used here is the step size
-    and there is no guarantee that a zero has been found. Consequently
-    the result should be verified. Safer algorithms are brentq,
-    brenth, ridder, and bisect, but they all require that the root
-    first be bracketed in an interval where the function changes
-    sign. The brentq algorithm is recommended for general use in one
-    dimensional problems when such an interval has been found.
+    sub-quadratic. This means that if the function is well behaved the
+    actual error in the estimated zero is approximately the square
+    (cube for Halley) of the requested tolerance up to roundoff error.
+
+    If `halt` is 'astep' or 'rstep', then the halting condition is
+    absolute or relative step size, and in this case there is no
+    guarantee that a zero has been found and the result should be
+    verified. If `halt` is changed to 'afval' or 'rfval' then the
+    absolute value of `func` or the relative size of `func` versus the
+    current iterate is used as a stopping criterion instead and an
+    error is raised if a root is not found. If the function is not
+    sufficiently well behaved, convergence to a root is still not
+    guaranteed.
+
+    Safer algorithms are `brentq`, `brenth`, `ridder`, and `bisect`,
+    but they all require that the root first be bracketed in an
+    interval where the function changes sign. The brentq algorithm is
+    recommended for general use in one dimensional problems when such
+    an interval has been found.
 
     """
+    if halt not in ['astep', 'rstep', 'afval', 'rfval']:
+        raise ValueError("invalid halting condition")
     if tol <= 0:
         raise ValueError("tol too small (%g <= 0)" % tol)
     if maxiter < 1:
         raise ValueError("maxiter must be greater than 0")
+
     if fprime is not None:
-        # Newton-Rapheson method
-        # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
+        # Newton-Raphson method
+        # Multiply by 1.0 to convert to floating point. We don't use float(x0)
         # so it still works if x0 is complex.
         p0 = 1.0 * x0
+        myargs = (p0,) + args
+        fval = func(*myargs)
```

Comment:
Indeed. But if the first function evaluation satisfies a termination condition (one of `*fval` for instance), there's no reason to perform another function evaluation. Also, upon first iteration two function evaluations are performed with only one check of termination condition so this is different from before (one more function evaluation) for no apparent reason as far as the default `astep` termination condition is concerned. So maybe this means that there cannot be a single point to check the termination condition for, but two, depending on whether `halt` is `*step` or `*fval`.